src/rust/ide/src/view/node_editor.rs

```diff
@@ -25,122 +92,293 @@ use weak_table::weak_value_hash_map::Entry::{Occupied, Vacant};
+            displayed_expressions,displayed_connections,logger});
+
         if let Err(err) = this.invalidate_graph() {
             error!(this.logger,"Error while initializing graph display: {err}");
         }
+
+        Self::setup_frp_network(&this);
         this
     }
 
+    fn setup_frp_network(this:&Rc<Self>) {
+        let network = &this.network;
+        let editor_outs = &this.editor.frp.outputs;
+        let weak    = Rc::downgrade(this);
+        let node_removed       = Self::define_action(this,Self::node_removed_action);
+        let connection_created = Self::define_action(this,Self::connection_created_action);
+        let connection_removed = Self::define_action(this,Self::connection_removed_action);
+        let node_moved         = Self::define_action(this,Self::node_moved_action);
+        frp::extend! {network
+            // Notifications from controller
+            def handle_notification = source::<Option<notification::Graph>>();
+            let (runner,is_action) = fence(&network,&handle_notification);
+            def _notification       = runner.map(f!([weak](notification) {
+                weak.upgrade().for_each(|this| this.handle_controller_notification(*notification));
+            }));
+            // Changes in Graph Editor
+            def _action = editor_outs.node_removed      .map2(&is_action,node_removed);
+            def _action = editor_outs.connection_added  .map2(&is_action,connection_created);
+            def _action = editor_outs.connection_removed.map2(&is_action,connection_removed);
+            def _action = editor_outs.node_position_set .map2(&is_action,node_moved);
+        }
+        Self::connect_frp_to_controller_notifications(this,handle_notification);
+    }
+
+    fn connect_frp_to_controller_notifications
+    (this:&Rc<Self>, frp_endpoint:frp::Source<Option<notification::Graph>>) {
+        let stream  = this.controller.graph.subscribe();
+        let weak    = Rc::downgrade(this);
+        let handler = process_stream_with_handle(stream,weak,move |notification,_this| {
+            frp_endpoint.emit_event(&Some(notification));
+            futures::future::ready(())
+        });
+        executor::global::spawn(handler);
+    }
+
+    fn define_action<Action,Parameter>
+    (this:&Rc<Self>, action:Action) -> impl Fn(&Parameter,&bool)
+    where Action : Fn(&Self,&Parameter) -> FallibleResult<()> + 'static {
+        let logger  = this.logger.clone_ref();
+        let weak    = Rc::downgrade(this);
+        move |parameter,is_action| {
+            if *is_action {
+                if let Some(this) = weak.upgrade() {
+                    let result = action(&*this,parameter);
+                    if let Err(err) = result {
+                        error!(logger,"Error while performing UI action on controllers: {err}");
+                        info!(logger,"Invlidating displayed graph");
```

Comment:
typo: Invlidating  => Invalidating 

But why is it conditionally logging that it is invalidating graph, if the condition body does nothing?